### DISCUSSION TITLE
Try to run Example1

### DISCUSSION THREAD

**Discussion by VishalPatel43** [LearnLib/learnlib#99] [General]
![Screenshot from 2023-11-01 01-55-43](https://github.com/LearnLib/learnlib/assets/123478469/1ddebb2e-f33e-498a-822c-c185aab83eba)

**Comment by mtf90** [comment]
I assume this is VSCode? Unfortunately, I am not too experienced with this IDE, so I cannot really comment on how to properly setup a project. But since LearnLib isn't any different from a regular Maven project, you should be able to find plenty of documentation.

Given that `GenerateRefinement` is an annotation from our annotation processor, maybe VSCode has problems with dependencies that are declared as `<scope>provided</scope>`. At least this one of the more plausible explanations that I can come up with. Maybe this gives you a direction you could look into.

**Comment by VishalPatel43** [comment]
No, this is intelliJ idea

The first time I was able to run this but after some time it led to the above error

Which version of Java should be preferred? currently, I am using Java 17


Till now I have tried Automatalib and Learnlib whose versions are currently in development

Now I am trying the stable version of Automatalib (0.10.0) and Learnlib (0.16.0)

**Reply by mtf90** [reply]
We aim at supporting the recents LTS releases of Java, which currently are 8, 11, 17, and 21.

If you are using the stable versions of AutomataLib/LearnLib, why do you have the code imported in Intellij? Typically, you would only import your own project and declare AutomataLib/LearnLib as a dependency. Then you also wouldn't need to compile the (already compiled) code of AutomataLib/LearnLib.

**Reply by VishalPatel43** [reply]
actually I imported that and just wants to run the given example code and understand the flow

I am not able to install automatalib 0.10.0 version library on my device

**Reply by mtf90** [reply]
How about you just create a new project from scratch that only uses AutomataLib/LearnLib as dependencies without importing the libraries themselves? You can use one of the existing archetypes for this. Simply run
```
mvn archetype:generate -DarchetypeGroupId=de.learnlib.archetypes -DarchetypeArtifactId=complete -DarchetypeVersion=0.16.0
```
and provide the project data as you like (e.g. `com.github.vishalpatel43` as groupId, etc.). This should give you a basic maven project that includes a simple example. You should be able to `mvn install` it, import it in IntelliJ and run it from IntelliJ. If this all works, you can start to expand the example to your liking.

**Comment by VishalPatel43** [comment]
I used this pom.xml file and successfully run the code which given examples in automatalib(0.10.0) and learnlib (0.16.0)

Thank you sir for your help, please tell me there will be changes in the POM.xml file

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>project</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <dependencies>
        <dependency>
            <groupId>net.automatalib.distribution</groupId>
            <artifactId>automata-distribution</artifactId>
            <version>0.10.0</version>
            <type>pom</type>
        </dependency>

        <dependency>
            <groupId>de.learnlib.distribution</groupId>
            <artifactId>learnlib-distribution</artifactId>
            <version>0.16.0</version>
            <type>pom</type>
        </dependency>
    </dependencies>

</project>
```

**Reply by mtf90** [reply]
Great thing you could get it running! Just one more note: LearnLib and AutomataLib occasionally use logging, but don't provide a logging implementation (we use the slf4j facade). So you might want to add an slf4j-compatible logging implementation if you are concerned about logging. Just add (for example)
```xml
<dependency>
    <groupId>ch.qos.logback</groupId>
    <artifactId>logback-classic</artifactId>
    <version>1.3.11</version>
</dependency>
```

As for changes, the distribution poms should remain identical in future releases. So even if the structure of the projects changes, you will always get the full package with the above dependencies. However, the code itself (package names, class names, etc.) might change of course, when you update the version. We usually track the most impactful changes in the respective changelogs ([here](https://github.com/LearnLib/automatalib/blob/develop/CHANGELOG.md) and [here](https://github.com/LearnLib/learnlib/blob/develop/CHANGELOG.md)).

**Reply by VishalPatel43** [reply]
Sir we discuss in #98 (https://github.com/LearnLib/learnlib/discussions/98)

Sir you provide the code for comparison between the **L*, NL*, KV, TTT** for time, membership queries and equivalence queries, number of states --> so what is the best input for these algo (same **SUL**)

Sir can you tell me how to generate random regular expression or random dfa or anything to provide these algo to the input and also provide how to verify the given input and resulted output will be same?
You provide the code which also avaialbe for stable version of the Learrnlib and Automatalib
Where I get the result of the all data time, membership queries and equivalence queries and many more

```


public static void main(String[] args) {
    final Automaton sul = new RegExp("a(b*|bb+)c?d").toAutomaton();
    final Alphabet<Character> alphabet = Alphabets.characters('a', 'd');
    final BricsDFA dfa = new BricsDFA(sul, true);
    final MembershipOracle<Character, Boolean > mqo = new SimulatorOracle<>(dfa);

    final CounterOracle<Character, Boolean> lstarCounter = new CounterOracle<>(mqo);
    final CounterOracle<Character, Boolean> nlstarCounter = new CounterOracle<>(mqo);

    final ClassicLStarDFA<Character> lstar = new ClassicLStarDFA<>(alphabet, lstarCounter);
    final NLStarLearner<Character> nlstar = new NLStarLearner<>(alphabet, nlstarCounter);

    // the challenge is to find an equivalence oracle that works for DFAs and NFAs
    // here, we simply use the W-method tests based on the determinized RegExp
    final SampleSetEQOracle<Character, Boolean> eqo = new SampleSetEQOracle<>(false);
    eqo.addAll(mqo, Lists.newArrayList(new WMethodTestsIterator<>(dfa, alphabet)));

    final Experiment<DFA<?, Character>> lstarExp = new Experiment<>(lstar, eqo, alphabet);
    final Experiment<NFA<?, Character>> nlstarExp = new Experiment<>(nlstar, eqo, alphabet);

    lstarExp.run();
    nlstarExp.run();

    System.out.println("LStar");
    System.out.println(lstarCounter.getStatisticalData().getDetails());
    System.out.println("NLStar");
    System.out.println(nlstarCounter.getStatisticalData().getDetails());

    Visualization.visualize(lstarExp.getFinalHypothesis(), alphabet);
    Visualization.visualize(nlstarExp.getFinalHypothesis(), alphabet);
}
```


dependency

```
<dependency>
    <groupId>dk.brics</groupId>
    <artifactId>automaton</artifactId>
</dependency>
```

**Reply by mtf90** [reply]
> Sir you provide the code for comparison between the L, NL, KV, TTT** for time, membership queries and equivalence queries, number of states --> so what is the best input for these algo (same SUL)

What do you mean with *inputs*? Configuration parameters like the `AcexAnalyzer`? It's hard to give a general answer because it often depends on the system that you are trying to learn. You may evaluate different strategies as part of your comparison or just stick with the default values from the respective `Builder` classes.

> Sir can you tell me how to generate random regular expression or random dfa or anything to provide these algo to the input and also provide how to verify the given input and resulted output will be same?

There are no direct utilities for generating (random) regular expressions in AutomataLib. As I mentioned earlier, the BRICS library may have some utilities for that. If you argue that regular expressions just describe regular languages and therefore any random automaton is fine for you, `RandomAutomata#randomDFA` might help you. However, I don't know if random regular expressions typically exhibit properties that plain random automata do not (e.g., loop-like structures when using Kleene-stars, etc.).

> You provide the code which also avaialbe for stable version of the Learrnlib and Automatalib

```java
final Automaton sul = new RegExp("a(b*|bb+)c?d").toAutomaton();
final Alphabet<Character> alphabet = Alphabets.characters('a', 'd');
final BricsDFA dfa = new BricsDFA(sul, true);
final MembershipOracle<Character, Boolean > mqo = new SimulatorOracle<>(dfa);

final JointCounterOracle<Character, Boolean> lstarCounter = new JointCounterOracle<>(mqo);
final JointCounterOracle<Character, Boolean> nlstarCounter = new JointCounterOracle<>(mqo);

final ClassicLStarDFA<Character> lstar = new ClassicLStarDFA<>(alphabet, lstarCounter);
final NLStarLearner<Character> nlstar = new NLStarLearner<>(alphabet, nlstarCounter);

// the challenge is to find an equivalence oracle that works for DFAs and NFAs
// here, we simply use the W-method tests based on the determinized RegExp
final SampleSetEQOracle<Character, Boolean> eqo = new SampleSetEQOracle<>(false);
eqo.addAll(mqo, Lists.newArrayList(new WMethodTestsIterator<>(dfa, alphabet, 0)));

final Experiment<DFA<?, Character>> lstarExp = new Experiment<>(lstar, eqo, alphabet);
final Experiment<NFA<?, Character>> nlstarExp = new Experiment<>(nlstar, eqo, alphabet);

lstarExp.run();
nlstarExp.run();

System.out.println("LStar");
System.out.println("Queries: " + lstarCounter.getQueryCount());
System.out.println("Symbols: " + lstarCounter.getSymbolCount());
System.out.println("NLStar");
System.out.println("Queries: " + nlstarCounter.getQueryCount());
System.out.println("Symbols: " + nlstarCounter.getSymbolCount());

Visualization.visualize(lstarExp.getFinalHypothesis(), alphabet);
Visualization.visualize(nlstarExp.getFinalHypothesis(), alphabet);
```

The main differences are that you need to use the old `JointCounterOracle` to track both the number of queries and symbols, and that the `WMethodTestsIterator` requires an additional parameter for the maximum exploration depth.

> Where I get the result of the all data time, membership queries and equivalence queries and many more

You need to track the duration for yourself (e.g., via `System.currentTimeMillis()`). For getting the number of equivalence queries you can query the `lstarExp.getRounds()` counter.

**Reply by VishalPatel43** [reply]
Then I will generate the Random DFAs uisng BRICS then give to SUL to the L*, NL*, KV and TTT, sir can your provide the code which generate the Random DFA and generate DFA which give to SUL in these algos and print the required value
I just need the start, I am using the stable version of the automatalib and learnlib

**Reply by mtf90** [reply]
Sorry, I haven't used the BRICS library to that level of detail either, so at this point we would both start from zero. And I think implementing your benchmark on top of a third-party library goes a little bit beyond the scope of giving support for AutomataLib/LearnLib. Maybe you can try to get in touch with the BRICS maintainers and ask for advice there.

**Reply by VishalPatel43** [reply]
Okay, Sir I  generated the random 1000 DFA which are in  dot format for my experiment, but can you provide that how to give SUL (oracle) to automata which is (.dot) file to the L* algortihm
please if possible, because learnlib and automatalib is very complex library and also I am trying to understand and explore these two  libraries and how to use properly

**Reply by mtf90** [reply]
Did you generate the automata with AutomataLib and then exported them to `.dot`? If so, you can simply use the in-memory automaton object and wrap it in a `SimulatorOracle` to pass it to a learning algorithm. If the `.dot` files are created externally, you may have a look at the [`DOTParsers`](https://github.com/LearnLib/automatalib/blob/develop/serialization/dot/src/main/java/net/automatalib/serialization/dot/DOTParsers.java) factory to import automata from `.dot` files. You may have to tinker around a little bit with the different node parsers and edge parsers depending on the structure of your `.dot` files (the parser-less methods typically only work with files that have been created by AutomataLib) but I think the semantics are quite straight-forward.

**Reply by VishalPatel43** [reply]
Sir I generate the DFA (.dot format) using the [AALpy](https://github.com/DES-Lab/AALpy/wiki) library.

.dot file of DFA

digraph random_dfa_1 {
s1 [label="s1"];
s2 [label="s2", shape=doublecircle];
s3 [label="s3", shape=doublecircle];
s4 [label="s4"];
s5 [label="s5", shape=doublecircle];
s6 [label="s6"];
s7 [label="s7"];
s1 -> s7  [label="a"];
s1 -> s6  [label="b"];
s2 -> s5  [label="a"];
s2 -> s4  [label="b"];
s3 -> s6  [label="a"];
s3 -> s6  [label="b"];
s4 -> s7  [label="a"];
s4 -> s2  [label="b"];
s5 -> s1  [label="a"];
s5 -> s4  [label="b"];
s6 -> s3  [label="a"];
s6 -> s2  [label="b"];
s7 -> s3  [label="a"];
s7 -> s5  [label="b"];
__start0 [label="", shape=none];
__start0 -> s1  [label=""];
}


Actually If there is any example of L* which take the dot file of DFA and learn it and it will give the result of the membership queries, equivalence queries, number state etc, then it will be helpful

Otherwise which format of DFA should I used how to generate the Random DFA using the Automatalib and then give to the SUL?

Learnlib and Automatalib libraries are large and little bit complex, I am also trying to learn this, but with your help it will be boost my progress and submission of my project is 8th Nov.

I learned the algortihms in my research (L*, NL*, KV, TTT) but its hard to use these libraries

It would be helpful if you provide the example with L* algorithms

**Reply by mtf90** [reply]
Unfortunately, there is no example for directly importing and learning a DFA from a `.dot` file. However, the example from [#98](https://github.com/LearnLib/learnlib/discussions/98#discussioncomment-7407786) covers the basic workflow of evaluating the learner performance. The only thing that has changed is how to construct the DFA/SUL that you pass to the simulator oracle.

For that purpose you can use the previously mentioned `DOTParsers` class. For the syntax of the AALpy export you are probably good to go with the default `#dfa` method. If you need to customize the node parsing, edge parsing, or initial nodes of the `.dot` file, you may use the more general `#fsa` method whose arguments describe how the properties of nodes and edges should be parsed. For your file, the `nodeParser` argument would receive (for example) the map `[("label", "s2"), ("shape", "doublecircle")]` and should return the respective state property (`true`/`false` for the acceptance of the state).

**Reply by mtf90** [reply]
Do you still need guidance, or can we mark this thread as resolved?
